<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Case Study: Benchmarking non-R Methods • SummarizedBenchmark</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Case Study: Benchmarking non-R Methods">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SummarizedBenchmark</a>
        <span class="version label label-info" data-toggle="tooltip" data-placement="bottom" title="Bioc Devel">2.3.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/SummarizedBenchmark-Introduction.html">Quick Start</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">General</li>
    <li>
      <a href="../articles/SummarizedBenchmark-Introduction.html">Introduction</a>
    </li>
    <li>
      <a href="../articles/SummarizedBenchmark-ClassDetails.html">Class Details</a>
    </li>
    <li>
      <a href="../articles/SummarizedBenchmark-FullCaseStudy.html">Full Case Study</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Features</li>
    <li>
      <a href="../articles/Feature-ErrorHandling.html">Error Handling</a>
    </li>
    <li>
      <a href="../articles/Feature-Iterative.html">Iterative Benchmarking</a>
    </li>
    <li>
      <a href="../articles/Feature-Parallel.html">Parallelization</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">More Case Studies</li>
    <li>
      <a href="../articles/CaseStudy-SingleCellSimulation.html">scRNA-seq Simulation</a>
    </li>
    <li>
      <a href="../articles/CaseStudy-RNAseqQuantification.html">RNA-seq Quantification</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/areyesq89/SummarizedBenchmark">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Case Study: Benchmarking non-R Methods</h1>
                        <h4 class="author">Alejandro Reyes, Patrick K. Kimes</h4>
            
            <h4 class="date">19 June 2019</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/areyesq89/SummarizedBenchmark/blob/master/vignettes/CaseStudy-RNAseqQuantification.Rmd"><code>vignettes/CaseStudy-RNAseqQuantification.Rmd</code></a></small>
      <div class="hidden name"><code>CaseStudy-RNAseqQuantification.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      “In this vignette, we provide an example of how the <em>SummarizedBenchmark</em> framework can be used to benchmark software tools that are not necessarily implemented in R. Note that the objetive of this vignette is <strong>not</strong> to exhaustively benchmark these methods, but rather demonstrate the usage of SummarizedBenchmark when the methods are not implemented in <em>R</em>.”
    </div>
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>To demonstrate the use of <em>SummarizedBenchmark</em> to compare software that is not written in <em>R</em>, we will compare the output of <em>sailfish</em>, <em>salmon</em> and <em>kallisto</em>, three alignment-free methods for transcript isoform quantification. Due to running time, disk space, and memory issues, some parts of this vignette are not run during the package build. Instead, we provide pre-computed objects containing the final <em>SummarizedBenchmark</em> object.</p>
</div>
<div id="data-preparation" class="section level1">
<h1 class="hasAnchor">
<a href="#data-preparation" class="anchor"></a>Data preparation</h1>
<p>We start by downloading the fastq files that we will use as input to quantify isoforms. We will use two samples from two brain replicates from the Mouse BodyMap <span class="citation">(Li et al. 2017)</span>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(BiocParallel)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/files2">dir.create</a></span>(<span class="st">"fastq"</span>, <span class="dt">showWarnings =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3"></a>
<a class="sourceLine" id="cb1-4" data-line-number="4">extractSRA &lt;-<span class="st"> </span><span class="cf">function</span>(sra_accession, <span class="dt">exe_path =</span> <span class="st">'fastq-dump'</span>, </a>
<a class="sourceLine" id="cb1-5" data-line-number="5">                       <span class="dt">args =</span> <span class="st">'--split-3 --gzip'</span>, <span class="dt">outdir =</span> <span class="st">'fastq'</span>, </a>
<a class="sourceLine" id="cb1-6" data-line-number="6">                       <span class="dt">dry_run =</span> <span class="ot">FALSE</span>) {</a>
<a class="sourceLine" id="cb1-7" data-line-number="7">    cmdline =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sprintf">sprintf</a></span>(<span class="st">'%s %s --outdir %s %s'</span>, </a>
<a class="sourceLine" id="cb1-8" data-line-number="8">                      exe_path, args, outdir, sra_accession)</a>
<a class="sourceLine" id="cb1-9" data-line-number="9">    <span class="cf">if</span> (dry_run) {</a>
<a class="sourceLine" id="cb1-10" data-line-number="10">        <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/message">message</a></span>(<span class="st">"will run with this command line:</span><span class="ch">\n</span><span class="st">"</span>, cmdline)</a>
<a class="sourceLine" id="cb1-11" data-line-number="11">    } <span class="cf">else</span> {</a>
<a class="sourceLine" id="cb1-12" data-line-number="12">        <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/function">return</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(cmdline))</a>
<a class="sourceLine" id="cb1-13" data-line-number="13">    }</a>
<a class="sourceLine" id="cb1-14" data-line-number="14">} </a>
<a class="sourceLine" id="cb1-15" data-line-number="15"></a>
<a class="sourceLine" id="cb1-16" data-line-number="16">samples &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"SRR5273705"</span>, <span class="st">"SRR5273689"</span>, <span class="st">"SRR5273699"</span>, <span class="st">"SRR5273683"</span>)</a>
<a class="sourceLine" id="cb1-17" data-line-number="17"></a>
<a class="sourceLine" id="cb1-18" data-line-number="18"><span class="kw"><a href="https://www.rdocumentation.org/packages/BiocParallel/topics/bplapply">bplapply</a></span>(samples, extractSRA, <span class="dt">BPPARAM =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/BiocParallel/topics/MulticoreParam-class">MulticoreParam</a></span>(<span class="dv">4</span>))</a>
<a class="sourceLine" id="cb1-19" data-line-number="19"></a>
<a class="sourceLine" id="cb1-20" data-line-number="20">annotation &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/data.frame">data.frame</a></span>(samples,</a>
<a class="sourceLine" id="cb1-21" data-line-number="21">                         <span class="dt">tissue =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"brain"</span>, <span class="st">"brain"</span>, <span class="st">"heart"</span>, <span class="st">"heart"</span>))</a></code></pre></div>
<p>Each of the three methods (<em>salmon</em>, <em>sailfish</em> and <em>kallisto</em>) require an indexing step for the reference transcriptome. We will use mouse annotations from the Gencode project. The code below downloads the mouse reference transcriptome.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/files2">dir.create</a></span>(<span class="st">"reference/raw"</span>, <span class="dt">recursive =</span> <span class="ot">TRUE</span>, <span class="dt">showWarnings =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/download.file">download.file</a></span>(<span class="st">"ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_mouse/release_M16/gencode.vM16.transcripts.fa.gz"</span>, </a>
<a class="sourceLine" id="cb2-3" data-line-number="3">              <span class="dt">destfile =</span> <span class="st">"reference/raw/transcriptome.fa.gz"</span>)</a></code></pre></div>
<p>Finally, we use the code below to build the transcriptome indices for the three different methods.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/files2">dir.create</a></span>(<span class="st">"reference/index"</span>, <span class="dt">showWarnings =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb3-2" data-line-number="2"></a>
<a class="sourceLine" id="cb3-3" data-line-number="3"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"kallisto index -i reference/index/kallistoIdx.idx reference/raw/transcriptome.fa.gz"</span>)</a>
<a class="sourceLine" id="cb3-4" data-line-number="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"salmon index -t reference/raw/transcriptome.fa.gz -i reference/index/salmon_index"</span>)</a>
<a class="sourceLine" id="cb3-5" data-line-number="5"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"gunzip -c reference/raw/transcriptome.fa.gz &gt; reference/raw/transcriptome.fa &amp;&amp; sailfish index -t reference/raw/transcriptome.fa -o reference/index/sailfish_index"</span>)</a>
<a class="sourceLine" id="cb3-6" data-line-number="6"></a>
<a class="sourceLine" id="cb3-7" data-line-number="7"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(Biostrings)</a>
<a class="sourceLine" id="cb3-8" data-line-number="8"></a>
<a class="sourceLine" id="cb3-9" data-line-number="9">dnSt &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/Biostrings/topics/XStringSet-io">readDNAStringSet</a></span>(<span class="st">"reference/raw/transcriptome.fa.gz"</span>))</a>
<a class="sourceLine" id="cb3-10" data-line-number="10">dnSt &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">sapply</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/strsplit">strsplit</a></span>(dnSt, <span class="st">"</span><span class="ch">\\</span><span class="st">||</span><span class="ch">\\</span><span class="st">."</span>), <span class="st">"[["</span>, <span class="dv">1</span>)</a></code></pre></div>
</div>
<div id="preparing-functions-with-system-calls-to-run-the-different-methods" class="section level1">
<h1 class="hasAnchor">
<a href="#preparing-functions-with-system-calls-to-run-the-different-methods" class="anchor"></a>Preparing functions with system calls to run the different methods</h1>
<p>If we want to use the <em>BenchDesign</em> infrastructure to compare tools that are run via the command line, we need to implement functions in <em>R</em> containing the system calls to the command line. Such functions must also collect the output of the methods and import them into <em>R</em>. To begin, we implement three functions that enable us to retrieve the version of the software that we will be running.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(SummarizedBenchmark)</a>
<a class="sourceLine" id="cb4-2" data-line-number="2"></a>
<a class="sourceLine" id="cb4-3" data-line-number="3">getKallistoVersion &lt;-<span class="st"> </span><span class="cf">function</span>() {</a>
<a class="sourceLine" id="cb4-4" data-line-number="4">    vers &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb4-5" data-line-number="5"><span class="st">        </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/warning">suppressWarnings</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"kallisto"</span>, <span class="dt">intern =</span> <span class="ot">TRUE</span>)[<span class="dv">1</span>])</a>
<a class="sourceLine" id="cb4-6" data-line-number="6">    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/strsplit">strsplit</a></span>(vers, <span class="st">" "</span>)[[<span class="dv">1</span>]][<span class="dv">2</span>]</a>
<a class="sourceLine" id="cb4-7" data-line-number="7">}</a>
<a class="sourceLine" id="cb4-8" data-line-number="8"></a>
<a class="sourceLine" id="cb4-9" data-line-number="9">getSalmonVersion &lt;-<span class="st"> </span><span class="cf">function</span>() {</a>
<a class="sourceLine" id="cb4-10" data-line-number="10">    vers &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb4-11" data-line-number="11"><span class="st">        </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/warning">suppressWarnings</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"salmon --version 2&gt;&amp;1"</span>, <span class="dt">intern =</span> <span class="ot">TRUE</span>)[<span class="dv">1</span>])</a>
<a class="sourceLine" id="cb4-12" data-line-number="12">    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/strsplit">strsplit</a></span>(vers, <span class="st">" "</span>)[[<span class="dv">1</span>]][<span class="dv">2</span>]</a>
<a class="sourceLine" id="cb4-13" data-line-number="13">}</a>
<a class="sourceLine" id="cb4-14" data-line-number="14"></a>
<a class="sourceLine" id="cb4-15" data-line-number="15">getSailfishVersion &lt;-<span class="st"> </span><span class="cf">function</span>() {</a>
<a class="sourceLine" id="cb4-16" data-line-number="16">    vers &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb4-17" data-line-number="17"><span class="st">        </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/warning">suppressWarnings</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(<span class="st">"sailfish --version 2&gt;&amp;1"</span>, <span class="dt">intern =</span> <span class="ot">TRUE</span>)[<span class="dv">1</span>])</a>
<a class="sourceLine" id="cb4-18" data-line-number="18">    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/strsplit">strsplit</a></span>(vers, <span class="st">" "</span>)[[<span class="dv">1</span>]][<span class="dv">3</span>]</a>
<a class="sourceLine" id="cb4-19" data-line-number="19">}</a></code></pre></div>
<p>Similarly, we can define <em>R</em> wrapper functions to run the different methods. Note that the functions below have three important characteristics. (1) They receive as input the arguments to the different methods such that <em>buildBench</em> can keep track of these, (2) they contain system calls that run the different methods and (3) they import the output of the different methods into <em>R</em> (in this case, using the <em><a href="https://bioconductor.org/packages/3.10/tximport">tximport</a></em> package).</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/files2">dir.create</a></span>(<span class="st">"out/kallisto"</span>, <span class="dt">showWarnings =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb5-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/files2">dir.create</a></span>(<span class="st">"out/salmon"</span>, <span class="dt">showWarnings =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb5-3" data-line-number="3"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/files2">dir.create</a></span>(<span class="st">"out/sailfish"</span>, <span class="dt">showWarnings =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb5-4" data-line-number="4"></a>
<a class="sourceLine" id="cb5-5" data-line-number="5">runKallisto &lt;-<span class="st"> </span><span class="cf">function</span>(sample, <span class="dt">args =</span> <span class="st">""</span>) {</a>
<a class="sourceLine" id="cb5-6" data-line-number="6">    fastqFile1 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sprintf">sprintf</a></span>( <span class="st">"fastq/%s_1.fastq.gz"</span>, sample)</a>
<a class="sourceLine" id="cb5-7" data-line-number="7">    fastqFile2 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/grep">gsub</a></span>( <span class="st">"_1"</span>, <span class="st">"_2"</span>, fastqFile1)</a>
<a class="sourceLine" id="cb5-8" data-line-number="8">    output &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sprintf">sprintf</a></span>(<span class="st">"out/kallisto/%s.out"</span>, sample)</a>
<a class="sourceLine" id="cb5-9" data-line-number="9">    cmd &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sprintf">sprintf</a></span>(<span class="st">"kallisto quant -i reference/index/kallistoIdx.idx -o %s %s %s %s"</span>, </a>
<a class="sourceLine" id="cb5-10" data-line-number="10">                   output, args, fastqFile1, fastqFile2)</a>
<a class="sourceLine" id="cb5-11" data-line-number="11">    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(cmd)</a>
<a class="sourceLine" id="cb5-12" data-line-number="12">    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">require</a></span>(tximport)</a>
<a class="sourceLine" id="cb5-13" data-line-number="13">    ab &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/tximport/topics/tximport">tximport</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/file.path">file.path</a></span>(output, <span class="st">"abundance.h5"</span>), </a>
<a class="sourceLine" id="cb5-14" data-line-number="14">                 <span class="dt">type =</span> <span class="st">"kallisto"</span>, <span class="dt">txOut =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb5-15" data-line-number="15">    counts &lt;-<span class="st"> </span>ab<span class="op">$</span>counts[, <span class="dv">1</span>]</a>
<a class="sourceLine" id="cb5-16" data-line-number="16">    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(counts) &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">sapply</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/strsplit">strsplit</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(counts), <span class="st">"</span><span class="ch">\\</span><span class="st">||</span><span class="ch">\\</span><span class="st">."</span>), <span class="st">"[["</span>, <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb5-17" data-line-number="17">    counts</a>
<a class="sourceLine" id="cb5-18" data-line-number="18">}</a>
<a class="sourceLine" id="cb5-19" data-line-number="19"></a>
<a class="sourceLine" id="cb5-20" data-line-number="20">runSalmon &lt;-<span class="st"> </span><span class="cf">function</span>(sample, <span class="dt">args =</span> <span class="st">"-l A -p 4"</span>) {</a>
<a class="sourceLine" id="cb5-21" data-line-number="21">    fastqFile1 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sprintf">sprintf</a></span>(<span class="st">"fastq/%s_1.fastq.gz"</span>, sample)</a>
<a class="sourceLine" id="cb5-22" data-line-number="22">    fastqFile2 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/grep">gsub</a></span>( <span class="st">"_1"</span>, <span class="st">"_2"</span>, fastqFile1)</a>
<a class="sourceLine" id="cb5-23" data-line-number="23">    output &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sprintf">sprintf</a></span>(<span class="st">"out/salmon/%s.out"</span>, sample)</a>
<a class="sourceLine" id="cb5-24" data-line-number="24">    cmd &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sprintf">sprintf</a></span>(<span class="st">"salmon quant -i reference/index/salmon_index %s -o %s -1 %s -2 %s"</span>,</a>
<a class="sourceLine" id="cb5-25" data-line-number="25">                   args, output, fastqFile1, fastqFile2)</a>
<a class="sourceLine" id="cb5-26" data-line-number="26">    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(cmd)</a>
<a class="sourceLine" id="cb5-27" data-line-number="27">    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">require</a></span>(tximport)</a>
<a class="sourceLine" id="cb5-28" data-line-number="28">    counts &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/tximport/topics/tximport">tximport</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/file.path">file.path</a></span>(output, <span class="st">"quant.sf"</span>), </a>
<a class="sourceLine" id="cb5-29" data-line-number="29">                       <span class="dt">type =</span> <span class="st">"salmon"</span>, <span class="dt">txOut =</span> <span class="ot">TRUE</span>)<span class="op">$</span>counts[, <span class="dv">1</span>]</a>
<a class="sourceLine" id="cb5-30" data-line-number="30">    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(counts) &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">sapply</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/strsplit">strsplit</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(counts), <span class="st">"</span><span class="ch">\\</span><span class="st">||</span><span class="ch">\\</span><span class="st">."</span>), <span class="st">"[["</span>, <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb5-31" data-line-number="31">    counts</a>
<a class="sourceLine" id="cb5-32" data-line-number="32">}</a>
<a class="sourceLine" id="cb5-33" data-line-number="33"></a>
<a class="sourceLine" id="cb5-34" data-line-number="34">runSailfish &lt;-<span class="st"> </span><span class="cf">function</span>(sample, <span class="dt">args =</span> <span class="st">"-l IU"</span>) {</a>
<a class="sourceLine" id="cb5-35" data-line-number="35">    fastqFile1 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sprintf">sprintf</a></span>( <span class="st">"fastq/%s_1.fastq.gz"</span>, sample)</a>
<a class="sourceLine" id="cb5-36" data-line-number="36">    fastqFile2 &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/grep">gsub</a></span>( <span class="st">"_1"</span>, <span class="st">"_2"</span>, fastqFile1)</a>
<a class="sourceLine" id="cb5-37" data-line-number="37">    output &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sprintf">sprintf</a></span>(<span class="st">"out/sailfish/%s.out"</span>, sample)</a>
<a class="sourceLine" id="cb5-38" data-line-number="38">    cmd &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sprintf">sprintf</a></span>(<span class="st">"echo </span><span class="ch">\"</span><span class="st">sailfish quant -i reference/index/sailfish_index %s -o %s -1 &lt;(zcat %s) -2 &lt;(zcat %s)</span><span class="ch">\"</span><span class="st"> | bash"</span>, </a>
<a class="sourceLine" id="cb5-39" data-line-number="39">                   args, output, fastqFile1, fastqFile2)</a>
<a class="sourceLine" id="cb5-40" data-line-number="40">    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/cat">cat</a></span>(cmd)</a>
<a class="sourceLine" id="cb5-41" data-line-number="41">    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system">system</a></span>(cmd)</a>
<a class="sourceLine" id="cb5-42" data-line-number="42">    counts &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/tximport/topics/tximport">tximport</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/file.path">file.path</a></span>(output, <span class="st">"quant.sf"</span>), </a>
<a class="sourceLine" id="cb5-43" data-line-number="43">                       <span class="dt">type =</span> <span class="st">"sailfish"</span>, <span class="dt">txOut =</span> <span class="ot">TRUE</span>)<span class="op">$</span>counts[, <span class="dv">1</span>]</a>
<a class="sourceLine" id="cb5-44" data-line-number="44">    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(counts) &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">sapply</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/strsplit">strsplit</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(counts), <span class="st">"</span><span class="ch">\\</span><span class="st">||</span><span class="ch">\\</span><span class="st">."</span>), <span class="st">"[["</span>, <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb5-45" data-line-number="45">    counts</a>
<a class="sourceLine" id="cb5-46" data-line-number="46">}</a></code></pre></div>
<p>Having defined these functions, we can now design our benchmark experiment using the <code><a href="../reference/BenchDesign.html">BenchDesign()</a></code> and <code><a href="../reference/addMethod.html">addMethod()</a></code> functions. For this specific experiment, we will run <em>salmon</em>, <em>sailfish</em> and <em>kallisto</em>. In addition, we will run <em>kallisto</em> and <em>salmon</em> both with default parameters and with their respective options to model for sequencing bias.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(SummarizedBenchmark)</a>
<a class="sourceLine" id="cb6-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(tximport)</a>
<a class="sourceLine" id="cb6-3" data-line-number="3"></a>
<a class="sourceLine" id="cb6-4" data-line-number="4">b &lt;-<span class="st"> </span><span class="kw"><a href="../reference/BenchDesign.html">BenchDesign</a></span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-5" data-line-number="5"><span class="st">    </span><span class="kw"><a href="../reference/addMethod.html">addMethod</a></span>(<span class="dt">label =</span> <span class="st">"kallisto-default"</span>,</a>
<a class="sourceLine" id="cb6-6" data-line-number="6">              <span class="dt">func =</span> runKallisto,</a>
<a class="sourceLine" id="cb6-7" data-line-number="7">              <span class="dt">params =</span> rlang<span class="op">::</span><span class="kw"><a href="https://rlang.r-lib.org/reference/quotation.html">quos</a></span>(<span class="dt">sample =</span> sample,</a>
<a class="sourceLine" id="cb6-8" data-line-number="8">                                   <span class="dt">args =</span> <span class="st">"-t 16"</span>),</a>
<a class="sourceLine" id="cb6-9" data-line-number="9">              <span class="dt">meta =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">pkg_vers =</span> rlang<span class="op">::</span><span class="kw"><a href="https://rlang.r-lib.org/reference/quotation.html">quo</a></span>(<span class="kw">getKallistoVersion</span>()))</a>
<a class="sourceLine" id="cb6-10" data-line-number="10">              ) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-11" data-line-number="11"><span class="st">    </span><span class="kw"><a href="../reference/addMethod.html">addMethod</a></span>(<span class="dt">label =</span> <span class="st">"kallisto-bias"</span>,</a>
<a class="sourceLine" id="cb6-12" data-line-number="12">             <span class="dt">func =</span> runKallisto,</a>
<a class="sourceLine" id="cb6-13" data-line-number="13">             <span class="dt">params =</span> rlang<span class="op">::</span><span class="kw"><a href="https://rlang.r-lib.org/reference/quotation.html">quos</a></span>(<span class="dt">sample =</span> sample,</a>
<a class="sourceLine" id="cb6-14" data-line-number="14">                                  <span class="dt">args =</span> <span class="st">"--bias -t 16"</span>),</a>
<a class="sourceLine" id="cb6-15" data-line-number="15">             <span class="dt">meta =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">pkg_vers =</span> rlang<span class="op">::</span><span class="kw"><a href="https://rlang.r-lib.org/reference/quotation.html">quo</a></span>(<span class="kw">getKallistoVersion</span>()))</a>
<a class="sourceLine" id="cb6-16" data-line-number="16">             ) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-17" data-line-number="17"><span class="st">   </span><span class="kw"><a href="../reference/addMethod.html">addMethod</a></span>(<span class="dt">label =</span> <span class="st">"salmon-default"</span>,</a>
<a class="sourceLine" id="cb6-18" data-line-number="18">             <span class="dt">func =</span> runSalmon,</a>
<a class="sourceLine" id="cb6-19" data-line-number="19">             <span class="dt">params =</span> rlang<span class="op">::</span><span class="kw"><a href="https://rlang.r-lib.org/reference/quotation.html">quos</a></span>(<span class="dt">sample =</span> sample,</a>
<a class="sourceLine" id="cb6-20" data-line-number="20">                                  <span class="dt">args =</span> <span class="st">"-l IU -p 16"</span>),</a>
<a class="sourceLine" id="cb6-21" data-line-number="21">             <span class="dt">meta =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">pkg_vers =</span> rlang<span class="op">::</span><span class="kw"><a href="https://rlang.r-lib.org/reference/quotation.html">quo</a></span>(<span class="kw">getSalmonVersion</span>()))</a>
<a class="sourceLine" id="cb6-22" data-line-number="22">             ) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-23" data-line-number="23"><span class="st">   </span><span class="kw"><a href="../reference/addMethod.html">addMethod</a></span>(<span class="dt">label =</span> <span class="st">"salmon-gcBias"</span>,</a>
<a class="sourceLine" id="cb6-24" data-line-number="24">             <span class="dt">func =</span> runSalmon,</a>
<a class="sourceLine" id="cb6-25" data-line-number="25">             <span class="dt">params =</span> rlang<span class="op">::</span><span class="kw"><a href="https://rlang.r-lib.org/reference/quotation.html">quos</a></span>(<span class="dt">sample=</span>sample,</a>
<a class="sourceLine" id="cb6-26" data-line-number="26">                                  <span class="dt">args=</span><span class="st">"-l IU --gcBias -p 16"</span>),</a>
<a class="sourceLine" id="cb6-27" data-line-number="27">             <span class="dt">meta =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">pkg_vers =</span> rlang<span class="op">::</span><span class="kw"><a href="https://rlang.r-lib.org/reference/quotation.html">quo</a></span>(<span class="kw">getSalmonVersion</span>()))</a>
<a class="sourceLine" id="cb6-28" data-line-number="28">             ) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb6-29" data-line-number="29"><span class="st">   </span><span class="kw"><a href="../reference/addMethod.html">addMethod</a></span>(<span class="dt">label =</span> <span class="st">"sailfish-default"</span>,</a>
<a class="sourceLine" id="cb6-30" data-line-number="30">             <span class="dt">func =</span> runSailfish,</a>
<a class="sourceLine" id="cb6-31" data-line-number="31">             <span class="dt">params =</span> rlang<span class="op">::</span><span class="kw"><a href="https://rlang.r-lib.org/reference/quotation.html">quos</a></span>(<span class="dt">sample=</span>sample,</a>
<a class="sourceLine" id="cb6-32" data-line-number="32">                                  <span class="dt">args=</span><span class="st">"-l IU -p 16"</span>),</a>
<a class="sourceLine" id="cb6-33" data-line-number="33">             <span class="dt">meta =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">pkg_vers =</span> rlang<span class="op">::</span><span class="kw"><a href="https://rlang.r-lib.org/reference/quotation.html">quo</a></span>(<span class="kw">getSailfishVersion</span>()))</a>
<a class="sourceLine" id="cb6-34" data-line-number="34">             )</a>
<a class="sourceLine" id="cb6-35" data-line-number="35"></a>
<a class="sourceLine" id="cb6-36" data-line-number="36"><span class="kw"><a href="../reference/printMethod.html">printMethods</a></span>(b)</a></code></pre></div>
<p>Now, the next step is to run the benchmark experiment. Since we are running the benchmark for two samples, we use an <code><a href="https://www.rdocumentation.org/packages/base/topics/lapply">lapply()</a></code> to loop over the sample names, run the benchmark experiment for each of them, and combine them using <code><a href="https://www.rdocumentation.org/packages/base/topics/cbind">cbind()</a></code>.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">dat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">txIDs =</span> dnSt)</a>
<a class="sourceLine" id="cb7-2" data-line-number="2">allSB &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">lapply</a></span>(samples, <span class="cf">function</span>(sample) {</a>
<a class="sourceLine" id="cb7-3" data-line-number="3">    dat[[<span class="st">"sample"</span>]] &lt;-<span class="st"> </span>sample</a>
<a class="sourceLine" id="cb7-4" data-line-number="4">    sb &lt;-<span class="st"> </span><span class="kw"><a href="../reference/buildBench.html">buildBench</a></span>(b, <span class="dt">data =</span> dat, <span class="dt">sortIDs =</span> <span class="st">"txIDs"</span>,</a>
<a class="sourceLine" id="cb7-5" data-line-number="5">                     <span class="dt">catchErrors =</span> <span class="ot">FALSE</span>, <span class="dt">parallel =</span> <span class="ot">TRUE</span>,</a>
<a class="sourceLine" id="cb7-6" data-line-number="6">                     <span class="dt">BPPARAM =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/BiocParallel/topics/SerialParam-class">SerialParam</a></span>())</a>
<a class="sourceLine" id="cb7-7" data-line-number="7">    <span class="kw">colData</span>( sb )<span class="op">$</span>sample &lt;-<span class="st"> </span>sample</a>
<a class="sourceLine" id="cb7-8" data-line-number="8">    <span class="kw">colData</span>( sb )<span class="op">$</span>tissue &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/character">as.character</a></span>(annotation<span class="op">$</span>tissue[annotation<span class="op">$</span>sample <span class="op">==</span><span class="st"> </span>sample])</a>
<a class="sourceLine" id="cb7-9" data-line-number="9">    sb</a>
<a class="sourceLine" id="cb7-10" data-line-number="10">})</a></code></pre></div>
<p>To keep the pre-computed object small, we will save the quantifications for only 50,000 randomly sampled transcripts.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1">keepRows &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">sapply</a></span>(allSB, <span class="cf">function</span>(x) { <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colSums">rowSums</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/NA">is.na</a></span>(<span class="kw">assay</span>(x))) })</a>
<a class="sourceLine" id="cb8-2" data-line-number="2">keepRows &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colSums">rowSums</a></span>(keepRows) <span class="op">==</span><span class="st"> </span><span class="dv">0</span></a>
<a class="sourceLine" id="cb8-3" data-line-number="3">allSB &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">lapply</a></span>(allSB, <span class="cf">function</span>(x) { x[keepRows, ] })</a>
<a class="sourceLine" id="cb8-4" data-line-number="4"></a>
<a class="sourceLine" id="cb8-5" data-line-number="5"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Random">set.seed</a></span>(<span class="dv">12</span>)</a>
<a class="sourceLine" id="cb8-6" data-line-number="6">keepRows &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sample">sample</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq_len</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(allSB[[<span class="dv">1</span>]])), <span class="dv">50000</span>)</a>
<a class="sourceLine" id="cb8-7" data-line-number="7"></a>
<a class="sourceLine" id="cb8-8" data-line-number="8">allSB &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">lapply</a></span>(allSB, <span class="cf">function</span>(x) { x[keepRows, ] })</a>
<a class="sourceLine" id="cb8-9" data-line-number="9"></a>
<a class="sourceLine" id="cb8-10" data-line-number="10"><span class="co">#save(allSB, file = "../data/quantSB.rda",</span></a>
<a class="sourceLine" id="cb8-11" data-line-number="11"><span class="co">#     compress = "xz", compression_level = 9)</span></a></code></pre></div>
<p>The resulting <code>allSB</code> object has been precomputed and can be loaded by doing</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(SummarizedBenchmark)</a>
<a class="sourceLine" id="cb9-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/data">data</a></span>(<span class="st">"quantSB"</span>)</a></code></pre></div>
<p>Notice that each element of the list contains a <em>SummarizedBenchmark</em> object in which the <code>colData()</code> contains both software versions and used parameters.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="kw">colData</span>(allSB[[<span class="dv">1</span>]])[, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">"param.args"</span>, <span class="st">"meta.version"</span>, <span class="st">"sample"</span>, <span class="st">"tissue"</span>)]</a></code></pre></div>
<pre><code>## DataFrame with 5 rows and 4 columns
##                              param.args meta.version      sample
##                             &lt;character&gt;  &lt;character&gt; &lt;character&gt;
## kallisto-default                "-t 16"       0.43.1  SRR5273705
## kallisto-bias            "--bias -t 16"       0.43.1  SRR5273705
## salmon-default            "-l IU -p 16"        0.9.1  SRR5273705
## salmon-gcBias    "-l IU --gcBias -p 16"        0.9.1  SRR5273705
## sailfish-default          "-l IU -p 16"       0.10.0  SRR5273705
##                       tissue
##                  &lt;character&gt;
## kallisto-default       brain
## kallisto-bias          brain
## salmon-default         brain
## salmon-gcBias          brain
## sailfish-default       brain</code></pre>
</div>
<div id="exploring-the-output-of-the-isoform-quantifications" class="section level1">
<h1 class="hasAnchor">
<a href="#exploring-the-output-of-the-isoform-quantifications" class="anchor"></a>Exploring the output of the isoform quantifications</h1>
<p>The code above returns a <em>SummarizedBenchmark</em> object containing the transcript isoform quantification results for the two samples. In this comparison, however, we don’t have a ground truth. However, having all the results in a single <em>SummarizedBenchmark</em> container facilitates the exploration of these results. For example, using a few lines of code, we can explore the similarity of the three methods using a standard dimensionality reduction technique (PCA).</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1">keep &lt;-<span class="st"> </span><span class="op">!</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colSums">rowSums</a></span>( <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/NA">is.na</a></span>( <span class="kw">assays</span>( allSB[[<span class="dv">1</span>]] )[[<span class="st">"default"</span>]] ) ) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span></a>
<a class="sourceLine" id="cb12-2" data-line-number="2"></a>
<a class="sourceLine" id="cb12-3" data-line-number="3">pcaRes &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb12-4" data-line-number="4"><span class="st">    </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/prcomp">prcomp</a></span>( <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">log10</a></span>( <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/t">t</a></span>( <span class="kw">assays</span>( allSB[[<span class="dv">1</span>]] )[[<span class="st">"default"</span>]][keep,] ) <span class="op">+</span><span class="st"> </span><span class="dv">1</span> ) )</a>
<a class="sourceLine" id="cb12-5" data-line-number="5">varExp &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Round">round</a></span>( <span class="dv">100</span><span class="op">*</span>(pcaRes<span class="op">$</span>sdev<span class="op">/</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sum">sum</a></span>( pcaRes<span class="op">$</span>sdev )), <span class="dv">2</span>)</a>
<a class="sourceLine" id="cb12-6" data-line-number="6"></a>
<a class="sourceLine" id="cb12-7" data-line-number="7">tidyData &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/data.frame">data.frame</a></span>( </a>
<a class="sourceLine" id="cb12-8" data-line-number="8">    <span class="dt">PC1=</span>pcaRes<span class="op">$</span>x[,<span class="st">"PC1"</span>], </a>
<a class="sourceLine" id="cb12-9" data-line-number="9">    <span class="dt">PC2=</span>pcaRes<span class="op">$</span>x[,<span class="st">"PC2"</span>], </a>
<a class="sourceLine" id="cb12-10" data-line-number="10">    <span class="dt">sample=</span><span class="kw">colData</span>( allSB[[<span class="dv">1</span>]] )<span class="op">$</span>sample, </a>
<a class="sourceLine" id="cb12-11" data-line-number="11">    <span class="dt">label=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/grep">gsub</a></span>(<span class="st">"</span><span class="ch">\\</span><span class="st">d+$"</span>, <span class="st">""</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>( <span class="kw">colData</span>(allSB[[<span class="dv">1</span>]]) ) ) )</a>
<a class="sourceLine" id="cb12-12" data-line-number="12"></a>
<a class="sourceLine" id="cb12-13" data-line-number="13">tidyData &lt;-<span class="st"> </span>tidyData <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb12-14" data-line-number="14"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span>( </a>
<a class="sourceLine" id="cb12-15" data-line-number="15">               <span class="dt">method=</span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/grep">gsub</a></span>( <span class="st">"-.*$"</span>, <span class="st">""</span>, label) )</a>
<a class="sourceLine" id="cb12-16" data-line-number="16"></a>
<a class="sourceLine" id="cb12-17" data-line-number="17"></a>
<a class="sourceLine" id="cb12-18" data-line-number="18">tidyData <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb12-19" data-line-number="19"><span class="st">    </span><span class="kw">ggplot</span>( <span class="kw">aes</span>( PC1, PC2, <span class="dt">colour=</span>label ) ) <span class="op">+</span></a>
<a class="sourceLine" id="cb12-20" data-line-number="20"><span class="st">    </span><span class="kw">geom_point</span>() <span class="op">+</span><span class="st"> </span><span class="kw">coord_fixed</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb12-21" data-line-number="21"><span class="st">    </span><span class="kw">ylab</span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sprintf">sprintf</a></span>( <span class="st">"PC2 (%0.2f %%)"</span>, varExp[<span class="dv">2</span>]) ) <span class="op">+</span></a>
<a class="sourceLine" id="cb12-22" data-line-number="22"><span class="st">    </span><span class="kw">xlab</span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sprintf">sprintf</a></span>( <span class="st">"PC1 (%0.2f %%)"</span>, varExp[<span class="dv">1</span>]) ) <span class="op">+</span></a>
<a class="sourceLine" id="cb12-23" data-line-number="23"><span class="st">    </span><span class="kw">theme</span>(<span class="dt">legend.pos=</span><span class="st">"top"</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb12-24" data-line-number="24"><span class="st">    </span><span class="kw">guides</span>(<span class="dt">col =</span> <span class="kw">guide_legend</span>(<span class="dt">nrow =</span> <span class="dv">5</span>), </a>
<a class="sourceLine" id="cb12-25" data-line-number="25">           <span class="dt">shape =</span> <span class="kw">guide_legend</span>(<span class="dt">nrow =</span> <span class="dv">4</span>))</a></code></pre></div>
<p><img src="CaseStudy-RNAseqQuantification_files/figure-html/pcaPlot-1.png" width="700"></p>
</div>
<div id="building-an-rnaseqcomp-object" class="section level1">
<h1 class="hasAnchor">
<a href="#building-an-rnaseqcomp-object" class="anchor"></a>Building an <em>rnaseqcomp</em> object</h1>
<p>For a more exhaustive benchmark of isoform quantification results, we refer the reader to the paper by <span class="citation">(Teng et al. 2016)</span>, which describes several metrics to evaluate the performance of isoform quantification pipelines. These metrics are implemented in <em><a href="https://bioconductor.org/packages/3.10/rnaseqcomp">rnaseqcomp</a></em>. The code below follows the steps from the <em><a href="https://bioconductor.org/packages/3.10/rnaseqcomp">rnaseqcomp</a></em> package to create an <em>rnaseqcomp</em> object. We recommend the readers to follow the vignette of the <em><a href="https://bioconductor.org/packages/3.10/rnaseqcomp">rnaseqcomp</a></em> package for an exhaustive benchmark of isoform quantification methods.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(rnaseqcomp)</a>
<a class="sourceLine" id="cb13-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(biomaRt)</a>
<a class="sourceLine" id="cb13-3" data-line-number="3"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/data">data</a></span>(simdata)</a>
<a class="sourceLine" id="cb13-4" data-line-number="4">houseHuman &lt;-<span class="st"> </span>simdata<span class="op">$</span>meta<span class="op">$</span>gene[simdata<span class="op">$</span>meta<span class="op">$</span>house]</a>
<a class="sourceLine" id="cb13-5" data-line-number="5">houseHuman &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/grep">gsub</a></span>(<span class="st">"</span><span class="ch">\\</span><span class="st">.</span><span class="ch">\\</span><span class="st">d+"</span>, <span class="st">""</span>, houseHuman )</a>
<a class="sourceLine" id="cb13-6" data-line-number="6"></a>
<a class="sourceLine" id="cb13-7" data-line-number="7">mart &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/biomaRt/topics/useMart">useMart</a></span>( <span class="st">"ensembl"</span>, <span class="st">"mmusculus_gene_ensembl"</span> )</a>
<a class="sourceLine" id="cb13-8" data-line-number="8">geneMap &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/biomaRt/topics/getBM">getBM</a></span>( <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>( <span class="st">"ensembl_transcript_id"</span>, </a>
<a class="sourceLine" id="cb13-9" data-line-number="9">                    <span class="st">"hsapiens_homolog_ensembl_gene"</span>, </a>
<a class="sourceLine" id="cb13-10" data-line-number="10">                    <span class="st">"hsapiens_homolog_orthology_type"</span> ), </a>
<a class="sourceLine" id="cb13-11" data-line-number="11">                 <span class="dt">mart =</span> mart)</a>
<a class="sourceLine" id="cb13-12" data-line-number="12">geneMap &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb13-13" data-line-number="13"><span class="st">    </span>geneMap[geneMap<span class="op">$</span><span class="st">`</span><span class="dt">hsapiens_homolog_orthology_type</span><span class="st">`</span> <span class="op">==</span><span class="st"> "ortholog_one2one"</span>,]</a>
<a class="sourceLine" id="cb13-14" data-line-number="14">houseMouse &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb13-15" data-line-number="15"><span class="st">    </span>geneMap<span class="op">$</span>ensembl_transcript_id[geneMap<span class="op">$</span>hsapiens_homolog_ensembl_gene <span class="op">%in%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb13-16" data-line-number="16"><span class="st">                                  </span>houseHuman]</a>
<a class="sourceLine" id="cb13-17" data-line-number="17"></a>
<a class="sourceLine" id="cb13-18" data-line-number="18">allSB &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/do.call">do.call</a></span>( cbind, allSB )</a>
<a class="sourceLine" id="cb13-19" data-line-number="19"><span class="kw">colData</span>( allSB )<span class="op">$</span>label &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/grep">gsub</a></span>(<span class="st">"</span><span class="ch">\\</span><span class="st">d+$"</span>, <span class="st">""</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>( <span class="kw">colData</span>( allSB ) ) )</a>
<a class="sourceLine" id="cb13-20" data-line-number="20"></a>
<a class="sourceLine" id="cb13-21" data-line-number="21">condInfo &lt;-<span class="st"> </span></a>
<a class="sourceLine" id="cb13-22" data-line-number="22"><span class="st">    </span><span class="kw">colData</span>( allSB )[<span class="kw">colData</span>( allSB )<span class="op">$</span>label <span class="op">==</span><span class="st"> "kallisto-default"</span>,<span class="st">"tissue"</span>]</a>
<a class="sourceLine" id="cb13-23" data-line-number="23"></a>
<a class="sourceLine" id="cb13-24" data-line-number="24">replicateInfo &lt;-<span class="st"> </span>condInfo</a>
<a class="sourceLine" id="cb13-25" data-line-number="25">evaluationFeature &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/rep">rep</a></span>( <span class="ot">TRUE</span>, <span class="dt">length.out =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>( allSB ) )</a>
<a class="sourceLine" id="cb13-26" data-line-number="26"></a>
<a class="sourceLine" id="cb13-27" data-line-number="27">calibrationFeature &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">rownames</a></span>(allSB) <span class="op">%in%</span><span class="st"> </span>houseMouse</a>
<a class="sourceLine" id="cb13-28" data-line-number="28"></a>
<a class="sourceLine" id="cb13-29" data-line-number="29">unitReference &lt;-<span class="st"> </span><span class="dv">1</span></a>
<a class="sourceLine" id="cb13-30" data-line-number="30">quantificationList &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">lapply</a></span>( </a>
<a class="sourceLine" id="cb13-31" data-line-number="31">    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/split">split</a></span>( <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq_along</a></span>( <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colnames">colnames</a></span>( allSB ) ), <span class="kw">colData</span>( allSB )<span class="op">$</span>label ), </a>
<a class="sourceLine" id="cb13-32" data-line-number="32">    <span class="cf">function</span>(x) { <span class="kw">assay</span>( allSB )[,x] })</a>
<a class="sourceLine" id="cb13-33" data-line-number="33"></a>
<a class="sourceLine" id="cb13-34" data-line-number="34">compObject &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/rnaseqcomp/topics/signalCalibrate">signalCalibrate</a></span>( </a>
<a class="sourceLine" id="cb13-35" data-line-number="35">    quantificationList, </a>
<a class="sourceLine" id="cb13-36" data-line-number="36">    <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/factor">factor</a></span>(condInfo), <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/factor">factor</a></span>(replicateInfo), </a>
<a class="sourceLine" id="cb13-37" data-line-number="37">    evaluationFeature, calibrationFeature, unitReference, </a>
<a class="sourceLine" id="cb13-38" data-line-number="38">    <span class="dt">calibrationFeature2 =</span> calibrationFeature)</a>
<a class="sourceLine" id="cb13-39" data-line-number="39"></a>
<a class="sourceLine" id="cb13-40" data-line-number="40">compObject</a></code></pre></div>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-Li_2017">
<p>Li, Bin, Tao Qing, Jinhang Zhu, Zhuo Wen, Ying Yu, Ryutaro Fukumura, Yuanting Zheng, Yoichi Gondo, and Leming Shi. 2017. “A Comprehensive Mouse Transcriptomic Bodymap Across 17 Tissues by Rna-Seq.” <em>Scientific Reports</em> 7 (1). <a href="https://doi.org/10.1038/s41598-017-04520-z" class="uri">https://doi.org/10.1038/s41598-017-04520-z</a>.</p>
</div>
<div id="ref-Teng_2016">
<p>Teng, Mingxiang, Michael I. Love, Carrie A. Davis, Sarah Djebali, Alexander Dobin, Brenton R. Graveley, Sheng Li, et al. 2016. “A Benchmark for Rna-Seq Quantification Pipelines.” <em>Genome Biology</em> 17 (1). <a href="https://doi.org/10.1186/s13059-016-0940-1" class="uri">https://doi.org/10.1186/s13059-016-0940-1</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#data-preparation">Data preparation</a></li>
      <li><a href="#preparing-functions-with-system-calls-to-run-the-different-methods">Preparing functions with system calls to run the different methods</a></li>
      <li><a href="#exploring-the-output-of-the-isoform-quantifications">Exploring the output of the isoform quantifications</a></li>
      <li><a href="#building-an-rnaseqcomp-object">Building an <em>rnaseqcomp</em> object</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="http://alejandroreyes.org">Alejandro Reyes</a>, <a href="https://www.pkimes.com">Patrick Kimes</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
