<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Case Study: scRNA-Seq Simulation • SummarizedBenchmark</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Case Study: scRNA-Seq Simulation">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SummarizedBenchmark</a>
        <span class="version label label-info" data-toggle="tooltip" data-placement="bottom" title="Bioc Devel">2.3.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/SummarizedBenchmark-Introduction.html">Quick Start</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">General</li>
    <li>
      <a href="../articles/SummarizedBenchmark-Introduction.html">Introduction</a>
    </li>
    <li>
      <a href="../articles/SummarizedBenchmark-ClassDetails.html">Class Details</a>
    </li>
    <li>
      <a href="../articles/SummarizedBenchmark-FullCaseStudy.html">Full Case Study</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Features</li>
    <li>
      <a href="../articles/Feature-ErrorHandling.html">Error Handling</a>
    </li>
    <li>
      <a href="../articles/Feature-Iterative.html">Iterative Benchmarking</a>
    </li>
    <li>
      <a href="../articles/Feature-Parallel.html">Parallelization</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">More Case Studies</li>
    <li>
      <a href="../articles/CaseStudy-SingleCellSimulation.html">scRNA-seq Simulation</a>
    </li>
    <li>
      <a href="../articles/CaseStudy-RNAseqQuantification.html">RNA-seq Quantification</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/areyesq89/SummarizedBenchmark">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Case Study: scRNA-Seq Simulation</h1>
                        <h4 class="author">Patrick K. Kimes, Alejandro Reyes</h4>
            
            <h4 class="date">22 June 2019</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/areyesq89/SummarizedBenchmark/blob/master/vignettes/CaseStudy-SingleCellSimulation.Rmd"><code>vignettes/CaseStudy-SingleCellSimulation.Rmd</code></a></small>
      <div class="hidden name"><code>CaseStudy-SingleCellSimulation.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      “In this vignette, we illustrate a simple approach to using the <em>SummarizedBenchmark</em> framework for organizing benchmarks with complex outputs, i.e. when methods return non-vector-like objects. This approach is demonstrated with a comparison of simulators for single-cell RNA-seq data implemented in the <em><a href="https://bioconductor.org/packages/3.10/splatter">splatter</a></em> package. SummarizedBenchmark package version: 2.3.4”
    </div>
    
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>Simulated data sets with known ground truths are often used for developing and comparing computational tools for genomic studies. However, the methods and approaches for simulating complex genomic data are rarely unified across studies. Recognizing this problem in the area of single-cell RNA-sequencing (scRNA-seq), the <em><a href="https://bioconductor.org/packages/3.10/splatter">splatter</a></em> package provides a uniform API for several “simulators” of scRNA-seq data, including the authors’ own “Splat” simulator <span class="citation">(Zappia, Phipson, and Oshlack 2017)</span>. In the <em><a href="https://bioconductor.org/packages/3.10/splatter">splatter</a></em> package, given a set of simulation parameters, each method returns a <em>SingleCellExperiment</em> object of simulated scRNA-seq counts.</p>
<p>Using comparisons presented in <span class="citation">(Zappia, Phipson, and Oshlack 2017)</span>, we illustrate how the <em>SummarizedBenchmark</em> framework can be used to perform comparisons when the output of each method is more complex than a vector of numbers (e.g. a <em>SingleCellExperiment</em>).</p>
</div>
<div id="building-the-benchdesign" class="section level1">
<h1 class="hasAnchor">
<a href="#building-the-benchdesign" class="anchor"></a>Building the BenchDesign</h1>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(<span class="st">"SummarizedBenchmark"</span>)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(<span class="st">"magrittr"</span>)</a></code></pre></div>
<p>Parameters for the simulators implemented in <em><a href="https://bioconductor.org/packages/3.10/splatter">splatter</a></em> can either be manually specified or estimated using existing data. Here, we use RSEM counts for a subset of high coverage samples in the <code>fluidigm</code> data set included in the <em><a href="https://bioconductor.org/packages/3.10/scRNAseq">scRNAseq</a></em> package. The data is made available as a <em>SummarizedExperiment</em> object.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(<span class="st">"splatter"</span>)</a>
<a class="sourceLine" id="cb2-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(<span class="st">"scRNAseq"</span>)</a>
<a class="sourceLine" id="cb2-3" data-line-number="3"></a>
<a class="sourceLine" id="cb2-4" data-line-number="4">fluidigm &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/scRNAseq/topics/ReprocessedData">ReprocessedFluidigmData</a></span>()</a>
<a class="sourceLine" id="cb2-5" data-line-number="5">se &lt;-<span class="st"> </span>fluidigm[, <span class="kw">colData</span>(fluidigm)[, <span class="st">"Coverage_Type"</span>] <span class="op">==</span><span class="st"> "High"</span>]</a>
<a class="sourceLine" id="cb2-6" data-line-number="6"><span class="kw">assays</span>(se) &lt;-<span class="st"> </span><span class="kw">assays</span>(se)[<span class="st">"rsem_counts"</span>]</a>
<a class="sourceLine" id="cb2-7" data-line-number="7"><span class="kw">assayNames</span>(se) &lt;-<span class="st"> "counts"</span></a></code></pre></div>
<p>For the purposes of this vignette, we only use a subset of the samples and genes.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Random">set.seed</a></span>(<span class="dv">1912</span>)</a>
<a class="sourceLine" id="cb3-2" data-line-number="2">se &lt;-<span class="st"> </span>se[<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sample">sample</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">nrow</a></span>(se), <span class="fl">1e4</span>), <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/sample">sample</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">ncol</a></span>(se), <span class="dv">20</span>)]</a></code></pre></div>
<p>To make comparisons with the simulated data sets easier, we convert the <em>SummarizedExperiment</em> object to the <em>SingleCellExperiment</em> class.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1">sce &lt;-<span class="st"> </span><span class="kw">as</span>(se, <span class="st">"SingleCellExperiment"</span>)</a></code></pre></div>
<p>Each of the simulators in the <em><a href="https://bioconductor.org/packages/3.10/splatter">splatter</a></em> package follow the <code>[prefix]Simulate</code> naming convention, with the corresponding parameter estimation function, <code>[prefix]Estimate</code>. Here, we use three methods included in the comparisons of <span class="citation">(Zappia, Phipson, and Oshlack 2017)</span>.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1">bd &lt;-<span class="st"> </span><span class="kw"><a href="../reference/BenchDesign.html">BenchDesign</a></span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-2" data-line-number="2"><span class="st">    </span><span class="kw"><a href="../reference/addMethod.html">addMethod</a></span>(<span class="dt">label =</span> <span class="st">"splat"</span>,</a>
<a class="sourceLine" id="cb5-3" data-line-number="3">              <span class="dt">func =</span> splatSimulate,</a>
<a class="sourceLine" id="cb5-4" data-line-number="4">              <span class="dt">params =</span> rlang<span class="op">::</span><span class="kw"><a href="https://rlang.r-lib.org/reference/quotation.html">quos</a></span>(<span class="dt">params =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/splatter/topics/splatEstimate">splatEstimate</a></span>(in_data),</a>
<a class="sourceLine" id="cb5-5" data-line-number="5">                                   <span class="dt">verbose =</span> in_verbose,</a>
<a class="sourceLine" id="cb5-6" data-line-number="6">                                   <span class="dt">seed =</span> in_seed)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-7" data-line-number="7"><span class="st">    </span><span class="kw"><a href="../reference/addMethod.html">addMethod</a></span>(<span class="dt">label =</span> <span class="st">"simple"</span>,</a>
<a class="sourceLine" id="cb5-8" data-line-number="8">              <span class="dt">func =</span> simpleSimulate,</a>
<a class="sourceLine" id="cb5-9" data-line-number="9">              <span class="dt">params =</span> rlang<span class="op">::</span><span class="kw"><a href="https://rlang.r-lib.org/reference/quotation.html">quos</a></span>(<span class="dt">params =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/splatter/topics/simpleEstimate">simpleEstimate</a></span>(in_data),</a>
<a class="sourceLine" id="cb5-10" data-line-number="10">                                   <span class="dt">verbose =</span> in_verbose,</a>
<a class="sourceLine" id="cb5-11" data-line-number="11">                                   <span class="dt">seed =</span> in_seed)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb5-12" data-line-number="12"><span class="st">    </span><span class="kw"><a href="../reference/addMethod.html">addMethod</a></span>(<span class="dt">label =</span> <span class="st">"lun"</span>,</a>
<a class="sourceLine" id="cb5-13" data-line-number="13">              <span class="dt">func =</span> lunSimulate,</a>
<a class="sourceLine" id="cb5-14" data-line-number="14">              <span class="dt">params =</span> rlang<span class="op">::</span><span class="kw"><a href="https://rlang.r-lib.org/reference/quotation.html">quos</a></span>(<span class="dt">params =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/splatter/topics/lunEstimate">lunEstimate</a></span>(in_data),</a>
<a class="sourceLine" id="cb5-15" data-line-number="15">                                   <span class="dt">verbose =</span> in_verbose,</a>
<a class="sourceLine" id="cb5-16" data-line-number="16">                                   <span class="dt">seed =</span> in_seed))</a></code></pre></div>
<p>Each simulator returns a single <em>SingleCellExperiment</em> object containing the simulated scRNA-seq counts. However, to fit the <em>SummarizedBenchmark</em> structure, each method in the <em>BenchDesign</em> must return a vector or a list. By default, if methods return objects which cannot be easily coerced to a matrix structure, the outputs from each assay is captured and combined using <code>rbind</code>. To prevent the warnings from printing, the user may specify <code>post = list</code> for all methods. However, note that this will result in all method output being wrapped in a list.</p>
</div>
<div id="running-the-benchmark-experiment" class="section level1">
<h1 class="hasAnchor">
<a href="#running-the-benchmark-experiment" class="anchor"></a>Running the Benchmark Experiment</h1>
<p>Using the <code>"counts"</code> assay of the <code>fluidigm</code> data set as input, we generate simulated data with the three methods.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1">fluidigm_dat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">in_data =</span> <span class="kw">assay</span>(sce, <span class="st">"counts"</span>),</a>
<a class="sourceLine" id="cb6-2" data-line-number="2">                     <span class="dt">in_verbose =</span> <span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb6-3" data-line-number="3">                     <span class="dt">in_seed =</span> <span class="dv">19120128</span>)</a>
<a class="sourceLine" id="cb6-4" data-line-number="4">sb &lt;-<span class="st"> </span><span class="kw"><a href="../reference/buildBench.html">buildBench</a></span>(bd, <span class="dt">data =</span> fluidigm_dat) </a></code></pre></div>
<pre><code>## Warning in FUN(X[[i]], ...): 
## Method outputs could not be reduced to matrix.</code></pre>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1">sb</a></code></pre></div>
<pre><code>## class: SummarizedBenchmark 
## dim: 1 3 
## metadata(1): sessions
## assays(1): default
## rownames: NULL
## rowData names(1): default
## colnames(3): splat simple lun
## colData names(7): func.pkg func.pkg.vers ... param.seed
##   session.idx</code></pre>
<p>The simulated data sets are returned as a single row in the assay of the <em>SummarizedBenchmark</em> object, with each column containing a list with a single <em>SingleCellExperiment</em> object.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="kw">assay</span>(sb)</a></code></pre></div>
<pre><code>##      splat simple lun
## [1,] ?     ?      ?</code></pre>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">sapply</a></span>(<span class="kw">assay</span>(sb), class)</a></code></pre></div>
<pre><code>## [1] "SingleCellExperiment" "SingleCellExperiment" "SingleCellExperiment"</code></pre>
</div>
<div id="comparing-the-results" class="section level1">
<h1 class="hasAnchor">
<a href="#comparing-the-results" class="anchor"></a>Comparing the Results</h1>
<p>Now that we have our set of simulated data sets, we can compare the behavior of each simulator. Fortunately, the <em><a href="https://bioconductor.org/packages/3.10/splatter">splatter</a></em> package includes two useful functions for comparing <em>SingleCellExperiment</em> objects (<code>compareSCEs</code> and <code>diffSCEs</code>). The assay of the <em>SummarizedBenchmark</em> can be passed directly to these functions. We also concatenate the original <code>fluidigm</code> data set, <code>sce</code>, with the simulated data sets for comparison.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1">res_compare &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/splatter/topics/compareSCEs">compareSCEs</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dt">ref =</span> sce, <span class="kw">assay</span>(sb)[<span class="dv">1</span>, ]))</a>
<a class="sourceLine" id="cb14-2" data-line-number="2">res_diff &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/splatter/topics/diffSCEs">diffSCEs</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dt">ref =</span> sce, <span class="kw">assay</span>(sb)[<span class="dv">1</span>, ]), <span class="dt">ref =</span> <span class="st">"ref"</span>)</a></code></pre></div>
<p>While these functions produce several metrics and plots, we only include two for illustration. More details on the output of these functions can be found in the documentation of the <em><a href="https://bioconductor.org/packages/3.10/splatter">splatter</a></em> package.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1">res_compare<span class="op">$</span>Plots<span class="op">$</span>MeanVar</a></code></pre></div>
<p><img src="CaseStudy-SingleCellSimulation_files/figure-html/plot-sim-result-comparison-1.png" width="700"></p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1">res_diff<span class="op">$</span>Plots<span class="op">$</span>MeanVar</a></code></pre></div>
<p><img src="CaseStudy-SingleCellSimulation_files/figure-html/plot-sim-result-comparison-2.png" width="700"></p>
<p>While (conveniently) functions already existed for comparing the simulated data sets, we can also define comparison metrics using the <em>SummarizedBenchmark</em> framework with <code>addPerformanceMetrics()</code>. We illustrate this feature using the “zeros per cell” and “zeros per gene” metrics shown in Figure 3 of <span class="citation">(Zappia, Phipson, and Oshlack 2017)</span>. Since the metric for each method is a vector (e.g. of zeros per cell) and not a single value, we again use <code>list</code> to wrap the output in the <code>evalFunction</code>.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1">sb &lt;-<span class="st"> </span>sb <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb17-2" data-line-number="2"><span class="st">    </span><span class="kw"><a href="../reference/addPerformanceMetric.html">addPerformanceMetric</a></span>(</a>
<a class="sourceLine" id="cb17-3" data-line-number="3">        <span class="dt">assay =</span> <span class="st">"default"</span>,</a>
<a class="sourceLine" id="cb17-4" data-line-number="4">        <span class="dt">evalMetric =</span> <span class="st">"zerosPerCell"</span>,</a>
<a class="sourceLine" id="cb17-5" data-line-number="5">        <span class="dt">evalFunction =</span> <span class="cf">function</span>(query, truth) {</a>
<a class="sourceLine" id="cb17-6" data-line-number="6">            <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colSums">colMeans</a></span>(<span class="kw">assay</span>(query[[<span class="dv">1</span>]], <span class="st">"counts"</span>) <span class="op">==</span><span class="st"> </span><span class="dv">0</span>)</a>
<a class="sourceLine" id="cb17-7" data-line-number="7">        }) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb17-8" data-line-number="8"><span class="st">    </span><span class="kw"><a href="../reference/addPerformanceMetric.html">addPerformanceMetric</a></span>(</a>
<a class="sourceLine" id="cb17-9" data-line-number="9">        <span class="dt">assay =</span> <span class="st">"default"</span>,</a>
<a class="sourceLine" id="cb17-10" data-line-number="10">        <span class="dt">evalMetric =</span> <span class="st">"zerosPerGene"</span>,</a>
<a class="sourceLine" id="cb17-11" data-line-number="11">        <span class="dt">evalFunction =</span> <span class="cf">function</span>(query, truth) {</a>
<a class="sourceLine" id="cb17-12" data-line-number="12">            <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/colSums">rowMeans</a></span>(<span class="kw">assay</span>(query[[<span class="dv">1</span>]], <span class="st">"counts"</span>) <span class="op">==</span><span class="st"> </span><span class="dv">0</span>)</a>
<a class="sourceLine" id="cb17-13" data-line-number="13">        } )</a></code></pre></div>
<p>Next, the metrics are calculated using <code>estimatePerformanceMetrics</code>. For plotting, we only keep the <code>label</code>, <code>value</code>, and <code>performanceMetric</code> columns of the returned table.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1">sbmets &lt;-<span class="st"> </span><span class="kw"><a href="../reference/estimateMetrics.html">estimatePerformanceMetrics</a></span>(sb, <span class="dt">tidy =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb18-2" data-line-number="2">sbmets  &lt;-<span class="st"> </span>dplyr<span class="op">::</span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span>(sbmets, label, value, performanceMetric)</a>
<a class="sourceLine" id="cb18-3" data-line-number="3"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(sbmets)</a></code></pre></div>
<pre><code>##    label        value performanceMetric
## 1  splat 0.606, 0....      zerosPerCell
## 2 simple 0.7076, ....      zerosPerCell
## 3    lun 0.3602, ....      zerosPerCell
## 4  splat 0.7, 0.4....      zerosPerGene
## 5 simple 1, 1, 0.....      zerosPerGene
## 6    lun 0.7, 0.4....      zerosPerGene</code></pre>
<p>Notice that the <code>value</code> is a list for each method and metric. These vectors can be expanded using <code><a href="https://tidyr.tidyverse.org/reference/unnest.html">tidyr::unnest</a></code>.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1">sbmets &lt;-<span class="st"> </span>tidyr<span class="op">::</span><span class="kw"><a href="https://tidyr.tidyverse.org/reference/unnest.html">unnest</a></span>(sbmets)</a>
<a class="sourceLine" id="cb20-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(sbmets)</a></code></pre></div>
<pre><code>##   label performanceMetric  value
## 1 splat      zerosPerCell 0.6060
## 2 splat      zerosPerCell 0.7024
## 3 splat      zerosPerCell 0.5881
## 4 splat      zerosPerCell 0.5717
## 5 splat      zerosPerCell 0.5932
## 6 splat      zerosPerCell 0.6351</code></pre>
<p>Finally, the performance metrics can be explored using standard plotting functions.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" data-line-number="1"><span class="kw">ggplot</span>(sbmets, <span class="kw">aes</span>(<span class="dt">x =</span> label, <span class="dt">y =</span> value,</a>
<a class="sourceLine" id="cb22-2" data-line-number="2">                   <span class="dt">color =</span> label, <span class="dt">fill =</span> label)) <span class="op">+</span></a>
<a class="sourceLine" id="cb22-3" data-line-number="3"><span class="st">    </span><span class="kw">geom_boxplot</span>(<span class="dt">alpha =</span> <span class="dv">1</span><span class="op">/</span><span class="dv">2</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb22-4" data-line-number="4"><span class="st">    </span><span class="kw">xlab</span>(<span class="st">"method"</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb22-5" data-line-number="5"><span class="st">    </span><span class="kw">scale_color_discrete</span>(<span class="st">"method"</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb22-6" data-line-number="6"><span class="st">    </span><span class="kw">scale_fill_discrete</span>(<span class="st">"method"</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb22-7" data-line-number="7"><span class="st">    </span><span class="kw">facet_grid</span>(performanceMetric <span class="op">~</span><span class="st"> </span>.) <span class="op">+</span></a>
<a class="sourceLine" id="cb22-8" data-line-number="8"><span class="st">    </span><span class="kw">theme_bw</span>()</a></code></pre></div>
<p><img src="CaseStudy-SingleCellSimulation_files/figure-html/plot-performace-metrics-1.png" width="700"></p>
</div>
<div id="new-data-sets" class="section level1">
<h1 class="hasAnchor">
<a href="#new-data-sets" class="anchor"></a>New Data Sets</h1>
<p>An advantage of the <em>SummarizedBenchmark</em> framework is that rerunning the comparisons with a new data set is as simple as calling <code>buildBench</code> with the same BenchDesign object paired with new <code>data =</code> input. To illustrate this, we run the same simulators again, but with simulation parameters estimated using the example single-cell count data included with the <em><a href="https://bioconductor.org/packages/3.10/scater">scater</a></em> package.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/data">data</a></span>(sc_example_counts, <span class="dt">package =</span> <span class="st">"scater"</span>)</a>
<a class="sourceLine" id="cb23-2" data-line-number="2">scec &lt;-<span class="st"> </span><span class="kw">SingleCellExperiment</span>(<span class="dt">assays =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">counts =</span> sc_example_counts))</a>
<a class="sourceLine" id="cb23-3" data-line-number="3"></a>
<a class="sourceLine" id="cb23-4" data-line-number="4">scater_dat &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/list">list</a></span>(<span class="dt">in_data =</span> scec,</a>
<a class="sourceLine" id="cb23-5" data-line-number="5">                   <span class="dt">in_verbose =</span> <span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb23-6" data-line-number="6">                   <span class="dt">in_seed =</span> <span class="dv">19120128</span>)</a>
<a class="sourceLine" id="cb23-7" data-line-number="7"><span class="kw"><a href="../reference/buildBench.html">buildBench</a></span>(bd, <span class="dt">data =</span> scater_dat)</a></code></pre></div>
<pre><code>## Warning in FUN(X[[i]], ...): 
## Method outputs could not be reduced to matrix.</code></pre>
<pre><code>## class: SummarizedBenchmark 
## dim: 1 3 
## metadata(1): sessions
## assays(1): default
## rownames: NULL
## rowData names(1): default
## colnames(3): splat simple lun
## colData names(7): func.pkg func.pkg.vers ... param.seed
##   session.idx</code></pre>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-Zappia_2017">
<p>Zappia, Luke, Belinda Phipson, and Alicia Oshlack. 2017. “Splatter: Simulation of Single-Cell Rna Sequencing Data.” <em>Genome Biology</em>. <a href="https://doi.org/10.1186/s13059-017-1305-0" class="uri">https://doi.org/10.1186/s13059-017-1305-0</a>.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#building-the-benchdesign">Building the BenchDesign</a></li>
      <li><a href="#running-the-benchmark-experiment">Running the Benchmark Experiment</a></li>
      <li><a href="#comparing-the-results">Comparing the Results</a></li>
      <li><a href="#new-data-sets">New Data Sets</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="http://alejandroreyes.org">Alejandro Reyes</a>, <a href="https://www.pkimes.com">Patrick Kimes</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
